src/modules/types.rs

```diff
@@ -14,14 +14,26 @@ pub enum Type {
     Generic
 }
 
+impl Type {
+    pub fn is_subset_of(&self, of: &Type) -> bool {
+        match (self, of) {
+            (Type::Array(l), Type::Array(r)) => {
+                **l != Type::Generic && **r == Type::Generic
+            }
+            (l, Type::Failable(r)) => !matches!(l, Type::Failable(_)) && (*l == **r || l.is_subset_of(r)),
+            _ => false
+        }
```

Comment:
Dont you need a base case for generic and non generic type? Then you can recursively call this función for arrays

Reply:
I don't understand what you mean. Generic arrays are Array that contain the Generic type, therefore I have to match on Array. I also cannot match on Array(Generic) right away, because it is inside a Box.

Reply:
I mean this:
```suggestion
        match (self, of) {
            (_, Type::Generic) => true,
            (Type::Array(l), Type::Array(r)) => l.is_subset_of(r),
            (l, Type::Failable(r)) => !matches!(l, Type::Failable(_)) && (*l == **r || l.is_subset_of(r)),
            _ => false
        }
```

Reply:
Yeah, even though nested arrays are not supported by Bash. But it makes more sense logically

Reply:
It does look a little bit nicer, but I don't think I want an unnecessary recursion call there. I don't think it will get inlined, so it will likely cost at runtime.

P.S. The other recursion call is kinda necessary and likely won't be called very often.